pkg/util/template/funcerror.go

```diff
@@ -0,0 +1,217 @@
+		if apiErrByReason != nil {
+			return apiErrByReason
+		}
+	}
+
+	return nil
+}
+
+func getErrorByReason(apiErr *apierrors.APIError, reasons []string) *apierrors.APIError {
+	for _, reason := range reasons {
+		if apiErr.Reason == reason {
+			return apiErr
+		}
+	}
+	return nil
+}
+
+// getValidationErrorByLocation returns an error with trimmed causes
+//
+// If all causes are trimmed, it returns nil
+// nolint: gocognit
+func getValidationErrorByLocation(apiErr *apierrors.APIError, locations []string) *apierrors.APIError {
+	out := apiErr.Clone()
+	causes, ok := out.Info["causes"]
+	if !ok {
+		panic("apiErr.Info['causes'] is not set")
+	}
+	causeArr, ok := causes.([]interface{})
+	if !ok {
+		panic("apiErr.Info['causes'] is not an array")
+	}
+	typedCauses := slice.Map(causeArr, func(cause interface{}) map[string]interface{} {
+		_cause, ok := cause.(map[string]interface{})
+		if !ok {
+			panic("apiErr.Info['causes'] is not an array of objects")
+		}
+		return _cause
+	})
+	trimmedCauses := trimValidationErrorCauses(typedCauses, locations)
+	if len(trimmedCauses) == 0 {
+		return nil
+	}
+	out.Info["causes"] = trimmedCauses
+	return out
+}
+
+// trimValidationErrorCauses removes causes that
+//
+//   - have mismatching location; AND
+//   - have details with ALL missing and expected fields that does not match required fields
+//     -- i.e. partially-matched cause will not be removed
+func trimValidationErrorCauses(causes []map[string]interface{}, locations []string) []map[string]interface{} {
+	trimmedCauses := slice.Map(causes, func(cause map[string]interface{}) map[string]interface{} {
+		causeLoc, ok := cause["location"]
+		if !ok {
+			panic("cause['location'] is not set")
+		}
+		causeLoc, ok = causeLoc.(string)
+		if !ok {
+			panic("cause['location'] is not string")
+		}
+		switch causeLoc {
+		case "":
+			return trimCauseWithoutLocation(cause, locations)
+		default:
+			return trimCauseWithLocation(cause, locations)
+		}
+	})
+
+	trimmedCausesWithoutNil := slice.Filter(trimmedCauses, func(cause map[string]interface{}) bool {
+		return cause != nil
+	})
+
+	return trimmedCausesWithoutNil
+}
+
+// trimCauseWithLocation returns matching location cause, otherwise nil
+func trimCauseWithLocation(cause map[string]interface{}, locations []string) map[string]interface{} {
+	for _, loc := range locations {
+		if cause["location"] == loc {
+			return cause
+		}
+	}
+	return nil
+}
+
+// trimCauseWithoutLocation removes causes with missing and expected fields that does not match required fields
+func trimCauseWithoutLocation(cause map[string]interface{}, requiredFields []string) map[string]interface{} {
+	kind, ok := cause["kind"]
+	if !ok {
+		panic("cause['kind'] is not set")
+	}
+	kind, ok = kind.(string)
+	if !ok {
+		panic("cause['kind'] is not string")
+	}
+	if kind != "required" {
+		panic(fmt.Errorf("unexpected cause without location - expected: required, got: %v", kind))
```

Comment:
Again, do not panic. We are handling error here. Panic in error handling will cause the original error to lose.

Instead we should handle error in a resilient way. I mention we only handle the case that we know, for unknown case, we just do nothing. See what I mean?

- We first check `kind`, instead of `location`.
- If `kind` is `required` and location is `""`, then this is a situation we know how to handle.
- Otherwise, if location is not `""`, then this is another situation we know how to handle.
- Otherwise, we do not panic and return gracefully.